src/librustc_codegen_ssa/back/link.rs

```diff
@@ -968,7 +968,80 @@ pub fn print_native_static_libs(sess: &Session, all_native_libs: &[NativeLibrary
     }
 }
 
+// Because windows-gnu target is meant to be self-contained for pure Rust code it bundles
+// own mingw-w64 libraries. These libraries are often not compatible with mingw-w64
+// installed in the system. This breaks many cases where Rust is mixed with other languages
+// (e.g. *-sys crates).
+// We prefer system mingw-w64 libraries if they are available to avoid this issue.
+fn get_crt_libs_path(sess: &Session) -> Option<PathBuf> {
+    fn find_exe_in_path<P>(exe_name: P) -> Option<PathBuf>
+    where
+        P: AsRef<Path>,
+    {
+        env::var_os("PATH").and_then(|paths| {
+            env::split_paths(&paths)
+                .filter_map(|dir| {
+                    let full_path = dir.join(&exe_name);
+                    if full_path.is_file() {
+                        Some(fix_windows_verbatim_for_gcc(&full_path))
+                    } else {
+                        None
+                    }
+                })
+                .next()
+        })
+    }
+
+    use std::sync::Mutex;
+    lazy_static::lazy_static! {
+        static ref SYSTEM_LIBS: Mutex<Option<Result<PathBuf, ()>>> = Mutex::new(None);
+    }
+
+    let system_libs = SYSTEM_LIBS.lock().unwrap().clone();
+    match system_libs {
+        Some(Ok(compiler_libs_path)) => Some(compiler_libs_path),
+        Some(Err(_)) => None,
+        _ => {
+            if let (linker, LinkerFlavor::Gcc) = linker_and_flavor(&sess) {
```

Comment:
Could the logic in this branch of the `match` be factored into its own function? Something like:

```
fn probe(sess: &Session) -> Option<PathBuf> { .. }
```

and that way this function would be a simple memoization of that `probe` function where we'd have `Mutex<Option<Option<PathBuf>>>` and if it's `None` we call `probe` and otherwise just clone/unwrap the result to return `Option<PathBuf>`